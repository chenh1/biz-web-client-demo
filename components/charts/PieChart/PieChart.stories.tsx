import type { PieChartProps } from './PieChart';
import React from 'react';
import { PieChart } from './PieChart';
import ParentSize from '@visx/responsive/lib/components/ParentSize';
import { Box } from '../../containers/Box/Box';

type Args = {
  args: PieChartProps
}

export const mockData = [
  { label: 'MSFT', shares: 360, value: 223.88 },
  { label: 'EL', shares: 123, value: 298.71 },
  { label: 'AMD', shares: 238, value: 92.12 },
  { label: 'V', shares: 291, value: 210.30 },
  { label: 'AAPL', shares: 663, value: 132.58 },
  { label: 'NKE', shares: 429, value: 143.75 },
  { label: 'WM', shares: 330, value: 118.23 }
]
export const totalShares = mockData.map(({ shares }) => shares).reduce((acc, prev) => acc + prev, 0)
export const totalValue = mockData.map(({ shares, value }) => shares * value).reduce((acc, prev) => acc + prev, 0)

const Template = (args: PieChartProps) => <Box className="min-h-[800px]"><ParentSize>{({ width, height }) => <PieChart {...args} width={width} height={height} />}</ParentSize></Box>

export default {
  component: Template
}

export const StandardPieChart: Args = {
  args: {
    height: 300,
    width: 500,
    data: mockData,
    pieValueOuter: d => (d.shares ?? 0) / totalShares,
    pieValueInner: d => (((d.value ?? 0) * (d.shares ?? 0)) / totalValue) * 100,
    pieValueOuterHighlightMap: ({ label, shares, ...rest }) => ({
      display: `${label}: ${shares} shares`,
      label,
      shares,
      ...rest
    }),
    pieValueInnerHighlightMap: ({ label, value, shares, ...rest }) => ({
      display: `${label}: $${Math.round(((shares ?? 0) * (value ?? 0)) * 100) / 100} total value`,
      label,
      shares,
      value,
      ...rest
    })
  }
};
