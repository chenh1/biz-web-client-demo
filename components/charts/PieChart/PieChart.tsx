import React, { useState } from 'react';
import Pie from '@visx/shape/lib/shapes/Pie';
import { DefaultOutput, scaleOrdinal } from '@visx/scale';
import { Group } from '@visx/group';
import { GradientPinkBlue } from '@visx/gradient';
import { animated, useTransition, to } from '@react-spring/web';
import { fontFamilyHeadings } from '../../constants/typography';
import { ScaleOrdinal } from '@visx/vendor/d3-scale';

interface PieChartData {
  label: string;
  display?: string;
  shares?: number;
  value?: number;
}

export type PieChartProps = {
  width: number;
  height: number;
  margin?: { top: number; right: number; bottom: number; left: number };
  animate?: boolean;
  pieValueOuter: (datum: PieChartData) => number;
  pieValueOuterHighlightMap: (datum: PieChartData) => PieChartData;
  pieValueInner: (datum: PieChartData) => number;
  pieValueInnerHighlightMap: (datum: PieChartData) => PieChartData;
  bgFill?: string;
  data: PieChartData[];
}

const getOuterSegmentColor: ScaleOrdinal<string, DefaultOutput> = scaleOrdinal<string>({
  domain: ['a', 'b', 'c', 'd'],
  range: [
    'rgba(255,255,255,0.7)',
    'rgba(255,255,255,0.6)',
    'rgba(255,255,255,0.5)',
    'rgba(255,255,255,0.4)',
    'rgba(255,255,255,0.3)',
    'rgba(255,255,255,0.2)',
    'rgba(255,255,255,0.1)',
  ],
});

const getInnerSegmentColor: ScaleOrdinal<string, DefaultOutput, never> = scaleOrdinal<string>({
  domain: ['a', 'b', 'c', 'd'],
  range: ['rgba(93,30,91,1)', 'rgba(93,30,91,0.8)', 'rgba(93,30,91,0.6)', 'rgba(93,30,91,0.4)'],
});

const defaultMargin = { top: 20, right: 20, bottom: 20, left: 20 };

export const PieChart: React.FC<PieChartProps> = ({
  width,
  height,
  margin = defaultMargin,
  animate = true,
  pieValueOuter,
  pieValueOuterHighlightMap,
  pieValueInner,
  pieValueInnerHighlightMap,
  bgFill = "url('#visx-pie-gradient')",
  data,
}) => {
  const [selectedOuterValue, setSelectedOuterValue] = useState<string | null>(null);
  const [selectedInnerValue, setSelectedInnerValue] = useState<string | null>(null);

  if (width < 10) return null;

  const innerWidth = width - margin.left - margin.right;
  const innerHeight = height - margin.top - margin.bottom;
  const radius = Math.min(innerWidth, innerHeight) / 2;
  const centerY = innerHeight / 2;
  const centerX = innerWidth / 2;
  const donutThickness = 50;

  return (
    <svg width={width} height={height}>
      <GradientPinkBlue id="visx-pie-gradient" />
      <rect rx={14} width={width} height={height} fill={bgFill} />
      <Group top={centerY + margin.top} left={centerX + margin.left}>
        <Pie
          data={
            selectedOuterValue
              ? data.filter(({ label }) => label === selectedOuterValue).map(pieValueOuterHighlightMap)
              : data.map(pieValueOuterHighlightMap)
          }
          pieValue={pieValueOuter}
          outerRadius={radius}
          innerRadius={radius - donutThickness}
          cornerRadius={3}
          padAngle={0.005}
        >
          {(pie) => (
            <AnimatedPie
              {...pie}
              animate={animate}
              getKey={(arc) => arc.data.display}
              onClickDatum={({ data: { label } }) =>
                animate &&
                setSelectedOuterValue(selectedOuterValue === label ? null : label)
              }
              getColor={(arc) => getOuterSegmentColor(arc.data.label) as string}
            />
          )}
        </Pie>
        <Pie
          data={
            selectedInnerValue
              ? data.filter(({ label }) => label === selectedInnerValue).map(pieValueInnerHighlightMap)
              : data.map(pieValueInnerHighlightMap)
          }
          pieValue={pieValueInner}
          pieSortValues={() => -1}
          outerRadius={radius - donutThickness * 1.3}
        >
          {(pie) => (
            <AnimatedPie
              {...pie}
              animate={animate}
              getKey={({ data: { display } }) => display}
              onClickDatum={({ data: { label } }) =>
                animate &&
                setSelectedInnerValue(selectedInnerValue === label ? null : label)
              }
              getColor={({ data: { label } }) => getInnerSegmentColor(label) as string}
            />
          )}
        </Pie>
      </Group>
      {animate && (
        <text
          textAnchor="end"
          x={width - 16}
          y={height - 16}
          fill="white"
          fontFamily={fontFamilyHeadings}
          fontSize={16}
          fontWeight={300}
          pointerEvents="none"
        >
          .
        </text>
      )}
    </svg>
  );
}

// Transition functions for the animated pie
const fromLeaveTransition = ({ endAngle }: { endAngle: number }) => ({
  startAngle: endAngle > Math.PI ? 2 * Math.PI : 0,
  endAngle: endAngle > Math.PI ? 2 * Math.PI : 0,
  opacity: 0,
});

const enterUpdateTransition = ({ startAngle, endAngle }: { startAngle: number; endAngle: number }) => ({
  startAngle,
  endAngle,
  opacity: 1,
});

// Define props for AnimatedPie
export type AnimatedPieProps = {
  animate: boolean;
  arcs: any[];
  path: any;
  getKey: (arc: any) => string;
  getColor: (arc: any) => string;
  onClickDatum: (arc: any) => void;
}

function AnimatedPie({
  animate,
  arcs,
  path,
  getKey,
  getColor,
  onClickDatum,
}: AnimatedPieProps) {
  const transitions = useTransition(arcs, {
    from: animate ? fromLeaveTransition : enterUpdateTransition,
    enter: enterUpdateTransition,
    update: enterUpdateTransition,
    leave: animate ? fromLeaveTransition : enterUpdateTransition,
    keys: getKey,
  });

  return transitions((props, arc, { key }) => {
    const [centroidX, centroidY] = path.centroid(arc);
    const hasSpaceForLabel = arc.endAngle - arc.startAngle >= 0.1;
    
    // @ts-ignore 
    const { startAngle, endAngle, opacity } = props || {}

    return (
      <g key={key}>
        <animated.path
          d={to([startAngle, endAngle], (startAngle, endAngle) =>
            path({
              ...arc,
              startAngle,
              endAngle,
            }),
          )}
          fill={getColor(arc)}
          onClick={() => onClickDatum(arc)}
          onTouchStart={() => onClickDatum(arc)}
        />
        {hasSpaceForLabel && (
          <animated.g style={{ opacity: opacity }}>
            <text
              fill="white"
              x={centroidX}
              y={centroidY}
              dy=".33em"
              fontSize={14}
              fontFamily={fontFamilyHeadings}
              textAnchor="middle"
              pointerEvents="none"
            >
              {getKey(arc)}
            </text>
          </animated.g>
        )}
      </g>
    );
  });
}
