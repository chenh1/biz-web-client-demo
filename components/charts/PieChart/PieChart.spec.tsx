import { PieChart } from "./PieChart";
import { render, waitFor } from '@testing-library/react';
import { mockData, totalShares, totalValue } from './PieChart.stories'
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window


it('renders a Pie chart given the dataset', async () => {
  const { asFragment } = render(
    <PieChart
      data={ mockData }
      height={300 }
      width={500 }
      pieValueOuter={d => (d.shares ?? 0) / totalShares }
      pieValueInner={d => (((d.value ?? 0) * (d.shares ?? 0)) / totalValue) * 100 }
      pieValueOuterHighlightMap={({ label, shares, ...rest }) => ({
        display: `${label}: ${shares} shares`,
        label,
        shares,
        ...rest
      })}
      pieValueInnerHighlightMap={({ label, value, shares, ...rest }) => ({
        display: `${label}: $${Math.round(((shares ?? 0) * (value ?? 0)) * 100) / 100} total value`,
        label,
        shares,
        value,
        ...rest
      })}
    />
  );

  await waitFor(() => {
    expect(asFragment()).toMatchSnapshot();
  });
});