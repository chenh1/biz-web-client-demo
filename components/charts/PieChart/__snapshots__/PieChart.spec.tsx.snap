// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Pie chart given the dataset 1`] = `
<DocumentFragment>
  <svg
    height="300"
    width="500"
  >
    <defs>
      <lineargradient
        id="visx-pie-gradient"
        x1="0"
        x2="0"
        y1="0"
        y2="1"
      >
        <stop
          offset="0%"
          stop-color="#F02FC2"
          stop-opacity="1"
        />
        <stop
          offset="100%"
          stop-color="#6094EA"
          stop-opacity="1"
        />
      </lineargradient>
    </defs>
    <rect
      fill="url('#visx-pie-gradient')"
      height="300"
      rx="14"
      width="500"
    />
    <g
      class="visx-group"
      transform="translate(250, 150)"
    >
      <g>
        <path
          d="M-2.3880612583373385e-14,-130L-1.4695761589768237e-14,-80Z"
          fill="rgba(255,255,255,0.3)"
        />
        <g
          style="opacity: 0;"
        >
          <text
            dy=".33em"
            fill="white"
            font-family=""Space Grotesk", sans-serif"
            font-size="14"
            pointer-events="none"
            text-anchor="middle"
            x="-14.255517846507757"
            y="104.02778576384243"
          >
            MSFT: 360 shares
          </text>
        </g>
      </g>
      <g>
        <path
          d="M-2.3880612583373385e-14,-130L-1.4695761589768237e-14,-80Z"
          fill="rgba(255,255,255,0.2)"
        />
        <g
          style="opacity: 0;"
        >
          <text
            dy=".33em"
            fill="white"
            font-family=""Space Grotesk", sans-serif"
            font-size="14"
            pointer-events="none"
            text-anchor="middle"
            x="-16.767097191966513"
            y="-103.65261430255943"
          >
            EL: 123 shares
          </text>
        </g>
      </g>
      <g>
        <path
          d="M-2.3880612583373385e-14,-130L-1.4695761589768237e-14,-80Z"
          fill="rgba(255,255,255,0.1)"
        />
        <g
          style="opacity: 0;"
        >
          <text
            dy=".33em"
            fill="white"
            font-family=""Space Grotesk", sans-serif"
            font-size="14"
            pointer-events="none"
            text-anchor="middle"
            x="-61.75194654045388"
            y="-84.92171158464086"
          >
            AMD: 238 shares
          </text>
        </g>
      </g>
      <g>
        <path
          d="M-2.3880612583373385e-14,-130L-1.4695761589768237e-14,-80Z"
          fill="rgba(255,255,255,0.7)"
        />
        <g
          style="opacity: 0;"
        >
          <text
            dy=".33em"
            fill="white"
            font-family=""Space Grotesk", sans-serif"
            font-size="14"
            pointer-events="none"
            text-anchor="middle"
            x="-101.52236037314161"
            y="-26.799446715668736"
          >
            V: 291 shares
          </text>
        </g>
      </g>
      <g>
        <path
          d="M7.960204194457796e-15,-130L4.898587196589413e-15,-80Z"
          fill="rgba(255,255,255,0.6)"
        />
        <g
          style="opacity: 0;"
        >
          <text
            dy=".33em"
            fill="white"
            font-family=""Space Grotesk", sans-serif"
            font-size="14"
            pointer-events="none"
            text-anchor="middle"
            x="79.1247850377109"
            y="-69.02367994200283"
          >
            AAPL: 663 shares
          </text>
        </g>
      </g>
      <g>
        <path
          d="M7.960204194457796e-15,-130L4.898587196589413e-15,-80Z"
          fill="rgba(255,255,255,0.5)"
        />
        <g
          style="opacity: 0;"
        >
          <text
            dy=".33em"
            fill="white"
            font-family=""Space Grotesk", sans-serif"
            font-size="14"
            pointer-events="none"
            text-anchor="middle"
            x="81.0286082929906"
            y="66.77847436188621"
          >
            NKE: 429 shares
          </text>
        </g>
      </g>
      <g>
        <path
          d="M-2.3880612583373385e-14,-130L-1.4695761589768237e-14,-80Z"
          fill="rgba(255,255,255,0.4)"
        />
        <g
          style="opacity: 0;"
        >
          <text
            dy=".33em"
            fill="white"
            font-family=""Space Grotesk", sans-serif"
            font-size="14"
            pointer-events="none"
            text-anchor="middle"
            x="-89.84391246666387"
            y="54.34216956179098"
          >
            WM: 330 shares
          </text>
        </g>
      </g>
      <g>
        <path
          d="M-1.1940306291686693e-14,-65L0,0Z"
          fill="rgba(93,30,91,1)"
        />
        <g
          style="opacity: 0;"
        >
          <text
            dy=".33em"
            fill="white"
            font-family=""Space Grotesk", sans-serif"
            font-size="14"
            pointer-events="none"
            text-anchor="middle"
            x="-19.690028446871068"
            y="-25.856387600773004"
          >
            MSFT: $80596.8 total value
          </text>
        </g>
      </g>
      <g>
        <path
          d="M-1.1940306291686693e-14,-65L0,0Z"
          fill="rgba(93,30,91,0.8)"
        />
        <g
          style="opacity: 0;"
        >
          <text
            dy=".33em"
            fill="white"
            font-family=""Space Grotesk", sans-serif"
            font-size="14"
            pointer-events="none"
            text-anchor="middle"
            x="-32.48766241530411"
            y="0.8954277130160033"
          >
            EL: $36741.33 total value
          </text>
        </g>
      </g>
      <g>
        <path
          d="M-1.1940306291686693e-14,-65L0,0Z"
          fill="rgba(93,30,91,0.6)"
        />
        <g
          style="opacity: 0;"
        >
          <text
            dy=".33em"
            fill="white"
            font-family=""Space Grotesk", sans-serif"
            font-size="14"
            pointer-events="none"
            text-anchor="middle"
            x="-28.50130749102901"
            y="15.618113564122066"
          >
            AMD: $21924.56 total value
          </text>
        </g>
      </g>
      <g>
        <path
          d="M-1.1940306291686693e-14,-65L0,0Z"
          fill="rgba(93,30,91,0.4)"
        />
        <g
          style="opacity: 0;"
        >
          <text
            dy=".33em"
            fill="white"
            font-family=""Space Grotesk", sans-serif"
            font-size="14"
            pointer-events="none"
            text-anchor="middle"
            x="-12.604766399343069"
            y="29.956132327421578"
          >
            V: $61197.3 total value
          </text>
        </g>
      </g>
      <g>
        <path
          d="M3.980102097228898e-15,-65L0,0Z"
          fill="rgba(93,30,91,1)"
        />
        <g
          style="opacity: 0;"
        >
          <text
            dy=".33em"
            fill="white"
            font-family=""Space Grotesk", sans-serif"
            font-size="14"
            pointer-events="none"
            text-anchor="middle"
            x="23.442745339502718"
            y="22.509724364088147"
          >
            AAPL: $87900.54 total value
          </text>
        </g>
      </g>
      <g>
        <path
          d="M3.980102097228898e-15,-65L0,0Z"
          fill="rgba(93,30,91,0.8)"
        />
        <g
          style="opacity: 0;"
        >
          <text
            dy=".33em"
            fill="white"
            font-family=""Space Grotesk", sans-serif"
            font-size="14"
            pointer-events="none"
            text-anchor="middle"
            x="29.367009541471404"
            y="-13.922239424429087"
          >
            NKE: $61668.75 total value
          </text>
        </g>
      </g>
      <g>
        <path
          d="M3.980102097228898e-15,-65L0,0Z"
          fill="rgba(93,30,91,0.6)"
        />
        <g
          style="opacity: 0;"
        >
          <text
            dy=".33em"
            fill="white"
            font-family=""Space Grotesk", sans-serif"
            font-size="14"
            pointer-events="none"
            text-anchor="middle"
            x="10.0708488744344"
            y="-30.900291308469946"
          >
            WM: $39015.9 total value
          </text>
        </g>
      </g>
    </g>
    <text
      fill="white"
      font-family=""Space Grotesk", sans-serif"
      font-size="16"
      font-weight="300"
      pointer-events="none"
      text-anchor="end"
      x="484"
      y="284"
    >
      .
    </text>
  </svg>
</DocumentFragment>
`;
